### ISSUE TITLE
https://test-videos.co.uk/vids/bigbuckbunny/mp4/h264/1080/Big_Buck_Bunny_1080_10s_30MB.mp4 doesn't play very well with openh264

### ISSUE THREAD

**Issue by jrmuizel** [cisco/openh264#3479]
I get errors like:
```
[libopenh264 @ 0x7fc59594f040] [OpenH264] this = 0x0x7fc59684c600, Warning:invalid ref pPic list reordering syntaxs!
[libopenh264 @ 0x7fc59594f040] DecodeFrame failed
[libopenh264 @ 0x7fc59594f040] [OpenH264] this = 0x0x7fc59684c600, Warning:referencing pictures lost due frame gaps exist, prev_frame_num: 14, curr_frame_num: 16
[libopenh264 @ 0x7fc59594f040] DecodeFrame failed
    Last message repeated 4 times
[libopenh264 @ 0x7fc59594f040] [OpenH264] this = 0x0x7fc59684c600, Warning:invalid ref pPic list reordering syntaxs!
[libopenh264 @ 0x7fc59594f040] DecodeFrame failed
```

**Comment by Darkspirit** [comment]
Flatpak Firefox (which uses OpenH264) just aborts playback in the middle of the video.
Other videos seem to have the same problem.

**Comment by akallabeth** [comment]
@jrmuizel Might be the same as #3476 and the one you already opened #3478

**Comment by xiaotiansf** [comment]
I am sure this issue is very different from #3476 and #3478. I will be looking at this issue next.
#3478 is also different from #3476. #3478's test case has b-frame while #3476's test case does not have b-frame.

**Comment by xiaotiansf** [comment]
@huili2 I trace down the issue with the extracted h264 elementary stream from Big_Buck_Bunny_1080_10s_30MB.mp4.
Found that all other errors starts with the error from 7th P_SLICE decoding in which uiMbType from the first MB block iMbXy = 0 returns wrong value. I attached the uiMbType comparison screen shot near the position. Left side is correct (from reference decoder), right side (from openh264 decoder) is incorrect. 
![Screenshot 2022-02-20 185530](https://user-images.githubusercontent.com/38150686/154881620-b9e2d0c9-307e-4e1b-922a-ca6d885b4f50.png).

There must have been a bug in P_ SLICE decoding.

**Comment by huili2** [comment]
thank @xiaotiansf for the detailed information. So let us check the issue then.

**Comment by huili2** [comment]
I tried with master (75b9fc) and v2.2.0 (33c6546), and both are OK to decode the internal 264 file without above error in windows. 
How to reproduce the error step by step? Thanks.

**Comment by xiaotiansf** [comment]
@huili2 
1. Extract h264 by ffmpeg -i Big_Buck_Bunny_1080_10s_30MB.mp4 -map 0:0 -vcodec copy -an Big_Buck_Bunny_1080_10s_30MB.264
2. Then just build openh264 console h264dec either on linux or windows on openh264 master.
I get consistent warning and errors as follows:
[OpenH264] this = 0x0x563198f19f10, Warning:invalid ref pPic list reordering syntaxs!
[OpenH264] this = 0x0x563198f19f10, Warning:referencing pictures lost due frame gaps exist, prev_frame_num: 14, curr_frame_num: 16
[OpenH264] this = 0x0x563198f19f10, Error:Ref Picture for B-Slice is lost, B-Slice decoding cannot be continued!
[OpenH264] this = 0x0x563198f19f10, Warning:DecodeCurrentAccessUnit() failed (394291) in frame: 17 uiDId: 0 uiQId: 0
[OpenH264] this = 0x0x563198f19f10, Error:Ref Picture for B-Slice is lost, B-Slice decoding cannot be continued!
[OpenH264] this = 0x0x563198f19f10, Warning:DecodeCurrentAccessUnit() failed (394291) in frame: 17 uiDId: 0 uiQId: 0
[OpenH264] this = 0x0x563198f19f10, Error:Ref Picture for B-Slice is lost, B-Slice decoding cannot be continued!
[OpenH264] this = 0x0x563198f19f10, Warning:DecodeCurrentAccessUnit() failed (394291) in frame: 17 uiDId: 0 uiQId: 0
[OpenH264] this = 0x0x563198f19f10, Error:Ref Picture for B-Slice is lost, B-Slice decoding cannot be continued!
[OpenH264] this = 0x0x563198f19f10, Warning:DecodeCurrentAccessUnit() failed (394291) in frame: 17 uiDId: 0 uiQId: 0

**Comment by huili2** [comment]
After some deep debugging, there's actually a bug in decoder in ref_pic_list_modification() module. Will fix it in recent days after a lot of tests.

**Comment by huili2** [comment]
@jrmuizel could you please check if current master has fixed the issue? Thanks.

**Comment by jrmuizel** [comment]
Yeah, playing on a faster computer confirms that decoding still doesn't match the builtin h264 decoder. I also confirmed this with:

`ffmpeg -i ../openh264/Big_Buck_Bunny_1080_10s_30MB.mp4 -vcodec ffv1 out1.mkv`

vs

`ffmpeg -vcodec libopenh264 -i ../openh264/Big_Buck_Bunny_1080_10s_30MB.mp4 -vcodec ffv1 out2.mkv`
